libs/parallelism/execution/include/hpx/execution/algorithms/on.hpp

```diff
@@ -7,88 +7,233 @@
+                            std::forward<Ts>(ts)...);
+                    }
+                };
+
+                template <typename E>
+                void set_error_predecessor_sender(E&& e) noexcept
+                {
+                    hpx::execution::experimental::set_error(
+                        std::move(receiver), std::forward<E>(e));
+                }
+
+                void set_done_predecessor_sender() noexcept
+                {
+                    hpx::execution::experimental::set_done(std::move(receiver));
+                }
+
+                template <typename... Us>
+                void set_value_predecessor_sender(Us&&... us) noexcept
+                {
+                    ts.template emplace<hpx::tuple<Us...>>(
+                        std::forward<Us>(us)...);
+                    scheduler_os.template emplace(
+                        hpx::util::detail::with_result_of([&]() {
+                            return hpx::execution::experimental::connect(
+                                hpx::execution::experimental::schedule(
+                                    std::move(scheduler)),
+                                scheduler_sender_receiver{*this});
+                        }));
+                    hpx::execution::experimental::start(scheduler_os.value());
+                }
+
+                struct scheduler_sender_receiver
+                {
+                    operation_state& os;
+
+                    template <typename E>
+                    void set_error(E&& e) noexcept
+                    {
+                        os.set_error_scheduler_sender(std::forward<E>(e));
+                    }
+
+                    void set_done() noexcept
+                    {
+                        os.set_done_scheduler_sender();
+                    };
+
+                    void set_value() noexcept
+                    {
+                        os.set_value_scheduler_sender();
+                    }
+                };
+
+                struct scheduler_sender_value_visitor
+                {
+                    std::decay_t<Receiver> receiver;
+
+                    void operator()(std::monostate) const
+                    {
+                        std::terminate();
+                    }
+
+                    template <typename Ts,
+                        typename = std::enable_if_t<
+                            !std::is_same_v<std::decay_t<Ts>, std::monostate>>>
+                    void operator()(Ts&& ts)
+                    {
+                        hpx::util::invoke_fused(
+                            hpx::util::bind_front(
+                                hpx::execution::experimental::set_value,
+                                std::move(receiver)),
+                            std::forward<Ts>(ts));
+                    }
+                };
+
+                template <typename E>
+                void set_error_scheduler_sender(E&& e) noexcept
```

Comment:
General question: would it be possible (make sense) to define all member functions on one-shot objects as `&&`?

Reply:
Possibly. The `connect`s on senders should already be marked `&&` where it makes sense. Operation state member functions could be marked `&`, but I'm not sure what it would gain us given that they should be non-copyable and non-movable. I don't really have a preference either way though. Do you think it would be preferable?

Reply:
Marking member functions as `&&` or `&` conveys intent, which is always a good thing as it might help reducing misuse. A good API is one that is easy to use correctly but difficult to use in unintended ways.

Reply:
Absolutely fair! I'll make the change at least for the P0443 APIs, and I'll see where it makes sense for the internal ones.